Fixes #29507 - Add puma-status for reporting

ehelms opened the pull request:
Replaces https://github.com/theforeman/foreman/pull/7578 with additional commit to ensure puma is 4.3.5 or higher.

sthirugn committed:
Fixes #29507 - Add puma-status for reporting

ehelms commented:
@tbrisker this would be nice to have in 2.1 for helping with debug

tbrisker commented:
tests are failing due to gemfile conflicts. Also, I assume this needs SELinux changes and packaging as well?

ehelms commented:
Packaging PR -- https://github.com/theforeman/foreman-packaging/pull/5031

I don't know if this requires SELinux changes, no issues were reported when testing. @lzap do you forsee any being needed?

lzap commented:
The only problem I can see is that the socket will have `tmp_t` (or similar) file label. If there's anything confined that needs to connect to it, it will require a rule to connect/read/write to `tmp_t`. Ideally this should have been a separate label.

But I believe `pumactl` runs unconfined (command line utility), so no issues there.

sthirugn commented:
Looks good, thank you for this PR.

ehelms commented:
Turns out I misread the puma repository structure and the 4.3.5 branch is missing the commit we need: 

https://github.com/puma/puma/commits/4.3.5

ares commented:
@ehelms this needs a rebase now, could you also please share an update about puma? There was some release already, so is this still blocked?

ehelms commented:
We never got a Puma 4 release with the needed security fix we implemented thus this has sat stagnant. A move to Puma 5 will enable this. Feel free to leave it open with that caveat or close it to be re-investigated at a later date.

ekohl commented:
Since https://projects.theforeman.org/issues/30490 `/run/foreman` should no longer be world readable so the security concern may no longer be valid.

ezr-ondrej commented:
@ehelms can we close this as Ewoud mentioned? I believe he's right and the concern is no longer valid.

ekohl commented:
> @ehelms can we close this as Ewoud mentioned? I believe he's right and the concern is no longer valid.

This is not what I meant this morning. I think that the blocker is no longer valid which means we can implement this. The feature itself is still valid.

ehelms commented:
Yep, this is valid and will be do-able with Puma 5 which we are actively working on.

ekohl commented:
> Yep, this is valid and will be do-able with Puma 5 which we are actively working on.

My argument was that the socket will live in `/run/foreman`. Currently that is mode 0700 so the security concern is no longer and issue so I don't see why we can't do this with Puma 4. With Puma 5 we can even better guarantee the security though.

ekohl commented:
ping @ehelms

ekohl commented:
https://github.com/theforeman/foreman/pull/8171 updates to Puma 5.1 which includes the fixes needed.

ehelms commented:
Intending to pick this up next sprint when I work on updating to Puma 5.1 across the board.

ezr-ondrej commented:
What is the plan for this PR? :)

ehelms commented:
Rebased and cleaned this up. Also opened  packaging PR to consume this https://github.com/theforeman/foreman-packaging/pull/6322

ekohl commented:
I think you meant to link https://github.com/theforeman/foreman-packaging/pull/6322. Is there also a related PR for Debian to maintain parity?

ehelms commented:
Deb link https://github.com/theforeman/foreman-packaging/pull/6323

ekohl commented on config/puma/production.rb:
```diff
@@ -1,3 +1,9 @@
+
+# Store server info state.
+state_path File.join(run_dir, 'puma.state')
+state_permission 0640
```
Looks like Rubocop is unhappy here

ekohl commented:
Looks like @ares wrote a tutorial for this: https://community.theforeman.org/t/puma-status-monitoring-not-working-at-the-moment/23154

ehelms commented:
I updated the Rubocop issue and rebased. Also updates the packaging PRs:

 * https://github.com/theforeman/foreman-packaging/pull/6322
 * https://github.com/theforeman/foreman-packaging/pull/6323

ekohl reviewed (approved):
This change looks correct. Probably good to get packaging changes approved too before merging.

evgeni commented:
RPM changes are good, Debian I need to look at, but I see no hurdles.
